wordpress-mobile/WordPress-Android
# UL&S: Improved login flow for passwordless accounts.

This PR introduces the following updates to the login flow:
- Users with passwords set now go straight to the password screen instead of being offered a magic link. On this password screen, there is a secondary option to choose a magic link.
- Users who don’t have passwords (a.k.a. passwordless users) will immediately be sent a magic link. This removes a separate confirmation step, and prevents us from asking passwordless users to enter a password.
- All instances of “magic link” in the user-facing text have been replaced with either “login link” or “signup link.”

Here's the updated flow diagram:

![continue-with-wordpress com](https://user-images.githubusercontent.com/830056/92644957-7ccefc00-f2ba-11ea-8300-75da52d31109.png)

Notes:
- When logging in with a passwordless account from the Email screen or tapping on the "Get a login link by email" button on the Password screen, the Magic Link Request screen will still show up for a moment (less than a second), but no interaction will be needed, as it will simply request the magic link automatically. Ideally, we would skip that screen entirely by requesting the magic link email and handling all possible responses from either the Email screen or the Password screen, but that would require a larger refactor that was considered out of scope for this PR. Besides, we already use similar approaches in some other cases, like when selecting a credential from Smart Lock, for example.
- This change still doesn't allow users to use magic links to connect social accounts (ref: #12216).

## To test

**Setup** (feel free to skip if you already have a passwordless account)

1. Log out if you are logged in.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is _not_ associated with a WordPress.com account.
1. Complete the sign-up flow normally.

**Test 1: Passwordless account / Magic Link login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a passwordless WordPress.com account.
1. Tap **Continue**.
1. Notice the Login Magic Link Sent screen.
1. Notice how there's no option to login with a password.
1. Tap **Check email**.
1. On the login email sent, tap **Log in to the app**.
1. Complete the login flow normally.

**Test 2: Account with password / Password login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a WordPress.com account that has a password set.
1. Tap **Continue**.
1. Notice the Password screen.
1. Notice how there's an option to login with a magic link.
1. Enter the WordPress.com account password.
1. Tap **Continue**.
1. Complete the login flow normally.

**Test 3: Account with password / Magic Link login**

0. Clear app data.
1. On the Prologue screen, if the Smart Lock dialog appears, dismiss it.
1. Tap **Continue with WordPress.com**.
1. On the Get Started screen, enter an email address that is associated with a WordPress.com account that has a password set.
1. Tap **Continue**.
1. Notice the Password screen.
1. Notice how there's an option to login with a magic link.
1. Tap **Get a login link by email**.
1. Notice the Login Magic Link Sent screen.
1. Tap **Check email**.
1. On the login email sent, tap **Log in to the app**.
1. Complete the login flow normally.

**Test 4: Secondary Entry Point / Account with password / Password login**

0. Clear app data.
1. From outside the app, try to share some content with WordPress.
1. If the Smart Lock dialog appears, dismiss it.
1. On the Get Started screen, enter an email address that is associated with a WordPress.com account that has a password set.
1. Tap **Continue**.
1. Notice the Password screen.
1. Notice how there's no option to login with a magic link.
1. Enter the WordPress.com account password.
1. Tap **Continue**.
1. Complete the login flow normally.
1. Notice how the Epilogue screen is not shown, and instead the user has to pick a site to share the content.

**Test 5: Secondary Entry Point / Passwordless account / Magic Link login**

0. Clear app data.
1. From outside the app, try to share some content with WordPress.
1. If the Smart Lock dialog appears, dismiss it.
1. On the Get Started screen, enter an email address that is associated with a passwordless WordPress.com account.
1. Tap **Continue**.
1. Notice the Login Magic Link Sent screen.
1. Notice how there's no option to login with a password.
1. Tap **Check email**.
1. On the login email sent, tap **Log in to the app**.
1. Complete the login flow normally.
1. Notice how the Epilogue screen is shown, and the user doesn't pick a site to share the content*.

\* This is the intended behavior here as it is preferable to interrupt the sharing flow than to prevent the user from logging in.

---

PR submission checklist:
